### DISCUSSION TITLE
Multiple deployments from single project / multiple projects with same repository

### DISCUSSION THREAD

**Discussion by yn5** [vercel/community#438] [Help] [answered]
We have a subfolder inside our repository that contains the styleguide of the project of which the build should be deployed separately from our Next.js project. We have it inside this repo since it renders the components from the Next.js project and having the components in a separate repository for both the Next.js project and the styleguide to use is effort and complexity we'd like to avoid.

As we didn't see ways to do multiple deployments (multiple build commandos, multiple output directories) it made sense to us to create a new project for the repository, but doing this results in a redirect to the existing Next.js project with the message "The project already exists.".

So we're wondering if there is a way for either one repository to be used in multiple projects. Or have one project deploy a subdirectory somehow?

Edit:
We also considered rendering the styleguide as pages in the Next.js site but we don't want them to be deployed in the production build.

**Comment by revskill10** [comment]
Vercel dashboard allows you to set subdirectory path to the project that you want to deploy. You can use it  (it's what's i'm doing for myproject with multiple clients)

**Reply by yn5** [reply]
Thanks for your extremely speedy reply! 🙂

The issue is that we also have the main Next.js project deployed on Vercel. So we need both to be deployed: the root directory as well as the subdirectory (where the root directory is a Next.js project and the subdirectory a static build). So we figured we could create a new project for the same repository where we indeed configure it to build and deploy that subdirectory, but that resulted in a redirect to the existing Next.js project with the message "The project already exists.".

**Comment by woutercouvaras** [comment]
The support guys are amazing, but I thought I'd share this here anyway - in case it saves someone some time.

My need is a bit different, but I think it falls under the same heading, so here goes...

I have one code base, which powers two different sites (i.e. two different vercel projects, each with their own ENVs, from the same repo).  

Previously one could run ```now -A sitename.now.json``` but since the new project linking functionality, this no longer works.

The way to do this in the current world is to find your teamId and the projectId and then to deploy each site using the following:

```VERCEL_ORG_ID=team_someId VERCEL_PROJECT_ID=someProjectId vercel --prod -A sitename.now.json```

Thanks to @paulogdm from support who pointed me to:  https://vercel.com/knowledge/using-vercel-cli-for-custom-workflows

**Reply by yn5** [reply]
Thanks for your help @woutercouvaras! Sorry for replying so late it took a while for me to find the time to try out your suggestions.

If I run the command you suggest and that can be found in the linked article the CLI just triggers a preview deployment of the existing project, adding the `--prod` option makes it a production deployment.

Using the CLI to redeploy the project without linking it to the existing project indeed links the codebase to a new project, but doesn't allow me to set up the integration with Github. If I try to do so (using the Vercel dashboard) I get the error `Error:The repository "<repository-name>" is already linked to the project "<project-name>".`

Is your `sitename.name.json` containing some configuration that allows specifying a different environment from the default Production, Preview and Development environments? Or is there another way to have a second project to be integrated with the same Github repository I'm missing maybe?

**Reply by woutercouvaras** [reply]
@yn5 Sorry for the delayed reply!

I see below that you got an answer from the Vercel support  team.  I never tried using the git integration because of my weird scenario, so I wouldn't have been able to shed any light.  I'm glad to hear they're working on this though.  They are constantly releasing amazing new features, which is fantastic.

All the best!

**Reply by alexisLefebvre** [reply]
Thanks! Using environment variables still work, we used it to deploy a website and storybook from the same code repository, by using 2 jobs in GitLab CI to push code to Vercel. Then we configured different commands in Vercel to compile the 2 websites.

**Comment by yn5** [comment]
For the future reader. I've had a very quick and nice conversation with Vercel's support team about the matter. As they where fine with me quoting the conversation here:

I asked:
> We would like to have multiple projects integrated with a single Github repository. As discussed here https://github.com/vercel/vercel/discussions/4719. What I understand from the replies in that discussion is that the only way to do that is to have the nth project deployed manually without a Github integration. Therefore we're considering setting up multiple Github repositories for the different Vercel projects we'd like and managing them as separate Git remotes. But before we do that I wanted to make sure there is no other way to do this that I'm missing maybe by asking via this route.

And they replied:
> Your understanding is correct, however, we are currently working on monorepo support that will hopefully be available in the next few weeks. This would allow you to import a repository an unlimited amount of times.

**Reply by Cocola** [reply]
Same concern for us! @yn5 any news about this ? Did you find a workaround ?

**Reply by paulogdm** [reply]
Monorepo support is a few days away. Stay tuned.

**Reply by johnomarkid** [reply]
@paulogdm checking in on monorepo support. It's a blocker for my company to use vercel right now.

**Reply by yn5** [reply]
> Same concern for us! @yn5 any news about this ? Did you find a workaround ?

No news on my side yet @Cocola.

> Monorepo support is a few days away. Stay tuned.

@paulogdm It's been a little while since you said this. Do you have any news to share for the time being maybe?

**Reply by paulogdm** [reply]
We are currently testing it under a feature flag: vercel.com/_flags.
It does require some polishing, which will take a couple of weeks, but you can already import multiple projects from a single repository.

After you enable the flag, try the import flow and let me know.

**Reply by yn5** [reply]
Thanks for your reply @paulogdm! The project I was hoping to use this for is in production so I’m a little hesitant to enable anything that says _Expirimental_ but I really appreciate the status update! (And will just be patient for a little longer.)

**Reply by Cocola** [reply]
@paulogdm Thanks for your reply ! The flag is enabled and we remove, re-import the project, nothing has changed in the settings tabs, did we miss something ?

**Reply by Cocola** [reply]
Sound good for us ! Thanks !

**Reply by colophonemes** [reply]
> We are currently testing it under a feature flag: vercel.com/_flags.

Is this feature flag still available? This is what I see when I view vercel.com/_flags. It's not obvious which of these flags corresponds to monorepo support.

![image](https://user-images.githubusercontent.com/6322017/91324037-7f5d2c00-e7b9-11ea-9327-22220ddb4257.png)

Also, assuming I do turn the flag on, how do I configure Vercel to build a specific project within the monorepo?

@woutercouvaras @paulogdm thanks very much for the pointer to `VERCEL_ORG_ID=team_someId VERCEL_PROJECT_ID=someProjectId vercel --prod -A sitename.now.json`, very useful in the meantime.

**Comment by yn5** [comment, marked as answer]
https://vercel.com/blog/monorepos 🥳

**Comment by makoto** [comment]
Hi, I am migrating from old `zeit now` based CLI tool to manage multiple env. Because `.vercel/project.json` only contains one project ID (in my case for my production env), how do I find project ID of other envs?

**Reply by woutercouvaras** [reply]
I used the api to find them:  https://vercel.com/docs/api#endpoints/projects

**Comment by adams-family** [comment]
I have two questions:

1) Is it possible to deploy a monorepo using `vercel-cli` without Git integration?

My point is that as sescribed in [https://vercel.com/blog/monorepos](https://vercel.com/blog/monorepos) all projects of a monorepo are automatically deployed on a Git push into the repo. Sometimes I only want to deploy let's say the UI but keep the API intact. I found the safest approach to do so is to deploy using the cli.

2) Has anyone successfully deployed a [NRWL](www.nrwl.io) monorepo (Angular + Express) to Vercel?

**Comment by younes0** [comment]
I have two differents Vercel projects with the same monorepo codebase.

![image](https://github.com/vercel/community/assets/886042/a1def7f0-2436-4018-993d-2b4a10d15159)

Root directory is `apps/backoffice` for backoffice project ,` apps/web` for web project
We disabled automatic deployment via `vercel.json` inside of each app (`web `/ `backoffice`) with :
```js
{
  "git": {
    "deploymentEnabled": {
      "main": false
    }
  }
}
```


When using `vercel deploy` from `apps` for the first time, I can only link one project and and therefore deploy this same project. 
Linking creates a `.vercel/project.json` (from root), so obviously only one project is handled there.
If I moved the `.vercel` folder to `apps/backoffice` for instance, an error occurs since the project root folder is not `apps/backoffice`.

So how can we deploy both projects manually ?

**Reply by amyegan** [reply]
Hey @younes0. Vercel CLI should always be invoked from the monorepo root, not a subdirectory. You can link to one Project at a time and use `vercel link` to switch to a different Project in the same monorepo. More info in the docs: [Using Monorepos with Vercel CLI](https://vercel.com/docs/monorepos#using-monorepos-with-vercel-cli)

This post is a few years old. If you're looking for more help with your monorepo, please [start a new discussion](https://github.com/vercel/community/discussions/new).
